File: Fenwick.cpp
template <typename T>
struct Fenwick {
    int n;
    std::vector<T> a;

    Fenwick(int n = 0) {
        init(n);
    }

    void init(int n) {
        this->n = n;
        a.assign(n, T());
    }

    void add(int x, T v) {
        for (int i = x; i <= n; i += i & -i) {
            a[i] += v;
        }
    }

    T sum(int x) {
        auto ans = T();
        for (int i = x; i > 0; i -= i & -i) {
            ans += a[i];
        }
        return ans;
    }

    T rangeSum(int l, int r) {
        return sum(r) - sum(l);
    }

    int kth(T k) {
        int x = 0;
        for (int i = 1 << std::__lg(n); i; i /= 2) {
            if (x + i <= n && k >= a[x + i - 1]) {
                x += i;
                k -= a[x - 1];
            }
        }
        return x;
    }
};

struct Max {
    int v;
    Max(int x = -1E9) : v{x} {}

    Max &operator+=(Max a) {
        v = std::max(v, a.v);
        return *this;
    }
};

struct Min {
    int v;
    Min(int x = 1e9) : v{x} {}

    Min &operator += (Min a) {
        v = min(v, a.v);
        return *this;
    }
};

File: go-template/Fenwick.go
package template

import (
	"math/bits"
)

type Fenwick struct {
	n int
	a []int
}

func NewFenwick(n int) *Fenwick {
	fenwick := &Fenwick{
		n: n,
		a: make([]int, n+1),
	}
	return fenwick
}

func (f *Fenwick) Add(x int, v int) {
	for ; x <= f.n; x += x & -x {
		f.a[x] += v
	}
}

func (f *Fenwick) Sum(x int) int {
	sum := 0
	for ; x > 0; x -= x & -x {
		sum += f.a[x]
	}
	return sum
}

func (f *Fenwick) RangeSum(l, r int) int {
	return f.Sum(r) - f.Sum(l-1)
}

func (f *Fenwick) Kth(k int) int {
	x := 0
	p := 1 << bits.Len(uint(f.n))
	for p > 0 {
		if x+p <= f.n && k >= f.a[x+p] {
			x += p
			k -= f.a[x]
		}
		p >>= 1
	}
	return x + 1
}
